stream-chat-android-offline/src/main/java/io/getstream/chat/android/livedata/controller/QueryChannelsController.kt

```diff
@@ -49,4 +63,28 @@ public interface QueryChannelsController {
      * If we are currently loading more channels
      */
     public val loadingMore: LiveData<Boolean>
+
+    public sealed class ChannelsState {
+        /** The QueryChannelsController is initialized but no query is currently running.
+         * If you know that a query will be started you typically want to display a loading icon.
+         * */
+        public object NoQueryActive : ChannelsState()
+        /** Indicates we are loading the first page of results.
+         * We are in this state if QueryChannelsController.loading is true
+         * For seeing if we're loading more results have a look at QueryChannelsController.loadingMore
+         *
+         * @see QueryChannelsController.loadingMore
+         * @see QueryChannelsController.loading
+         * */
+        public object Loading : ChannelsState()
+        /** If we are offline and don't have channels stored in offline storage, typically displayed as an error condition. */
+        public object OfflineNoResults : ChannelsState()
+        /** If we have no offline data and get an error from the API */
+        public class Failed(public val error: ChatError) : ChannelsState()
```

Comment:
This `Failed` state also seems unused at the moment

Reply:
We should trigger the failed state when you hit an API error and no local storage results are available. This is important to get right since it can be very confusing if we show things like an infinite loading icon when there is in fact an error.

Reply:
@JcMinarro ^

Reply:
Failure state is not needed in our current implementation.
We have 2 ways to obtain data, remote datasource (Stream Server) and local datasource (DB) that are run in parallel. If both of them fail, we will obtain an `OfflineNoResults` state, and the query will be marked as "recovery needed", so no more infinite loading state could be there. If a controller is marked as "recovery needed" (because a query failed) it is because there was a network connection failure, when the connection is live again or any other events come to the app, the query will be processed again. Because we are using a react approach from our DB, any new message/channel is inserted in our DB will trigger an update in our controller